src/pkg/packager/lint/lint.go

```diff
@@ -26,119 +24,99 @@ import (
 )
 
 // ZarfSchema is exported so main.go can embed the schema file
-var ZarfSchema embed.FS
+var ZarfSchema fs.ReadFileFS
 
-func getSchemaFile() ([]byte, error) {
-	return ZarfSchema.ReadFile("zarf.schema.json")
-}
-
-// Validate validates a zarf file against the zarf schema, returns *validator with warnings or errors if they exist
-// along with an error if the validation itself failed
-func Validate(ctx context.Context, createOpts types.ZarfCreateOptions) (*Validator, error) {
-	validator := Validator{}
-	var err error
-
-	if err := utils.ReadYaml(filepath.Join(createOpts.BaseDir, layout.ZarfYAML), &validator.typedZarfPackage); err != nil {
+// Validate the given Zarf package. The Zarf package should not already be composed when sent to this function.
+func Validate(ctx context.Context, pkg types.ZarfPackage, createOpts types.ZarfCreateOptions) ([]types.PackageFinding, error) {
+	var findings []types.PackageFinding
+	compFindings, err := lintComponents(ctx, pkg, createOpts)
+	if err != nil {
 		return nil, err
 	}
+	findings = append(findings, compFindings...)
 
-	if err := utils.ReadYaml(filepath.Join(createOpts.BaseDir, layout.ZarfYAML), &validator.untypedZarfPackage); err != nil {
+	jsonSchema, err := ZarfSchema.ReadFile("zarf.schema.json")
+	if err != nil {
 		return nil, err
 	}
 
-	if err := os.Chdir(createOpts.BaseDir); err != nil {
-		return nil, fmt.Errorf("unable to access directory '%s': %w", createOpts.BaseDir, err)
-	}
-
-	validator.baseDir = createOpts.BaseDir
-
-	lintComponents(ctx, &validator, &createOpts)
-
-	if validator.jsonSchema, err = getSchemaFile(); err != nil {
+	var untypedZarfPackage interface{}
+	if err := utils.ReadYaml(layout.ZarfYAML, &untypedZarfPackage); err != nil {
 		return nil, err
 	}
```

Comment:
Why do we read the zarf.yaml into an interface here? Can we just pass the `pkg` argument to `validateSchema` instead?

Reply:
Good question, but it needs to be untyped because otherwise we would lose certain data when the yaml values are put into a struct and given zero values. The TestZarfSchema function in this PR has examples of where this is applicable (github is being weird about giving me a link)

Reply:
I deleted this code and `TestZarfSchema` still passes because `runSchema` accepts an interface, so we can still pass an arbitrary, untyped object in the test without having to read the file into memory again in `Validate`. So our actual code is strongly typed, but untyped for testing invalid schema fields and what not

Reply:
You can still pass in the package but when a yaml file is read into a Go struct it will drop fields that are not on the yaml object and Go will give zero values to every non pointer field in the struct. For example, you cannot replicate the below errors if you pass in a `ZarfPackage` struct. The extra field will be dropped, you can't make a string an int, and the `name` field will pass the required check since it exists with a `""` value . You can also see this by running the e2e test `TestLint`
![image](https://github.com/defenseunicorns/zarf/assets/37223396/071a9524-130c-49a8-b2a8-149e92513d01)

Reply:
Oh I understand what you're saying now after looking at the `TestLint` e2e test. Makes sense thanks for the explanation